File: 백준/Silver/1992. 쿼드트리/쿼드트리.py
import sys

def quad_tree(graph):
    n = len(graph)
    if n == 1:
        return graph[0][0]
    tmp = []
    if sum([sum(list(map(int,list(i)))) for i in graph]) == n**2 or sum([sum(list(map(int,list(i)))) for i in graph]) == 0:
        return graph[0][0]

    else:
        string_return = "("
        for i in range(2):
            for j in range(2):
                string_return += quad_tree([dots[n//2*j:n//2*(j+1)] for dots in graph[n//2*i:n//2*(i+1)]])
                
        string_return += ")"
        return string_return

N = int(input())
graph = [input()for _ in range(N)]
print(quad_tree(graph))

File: 백준/Silver/1992. 쿼드트리/쿼드트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
	static StringBuilder sb = new StringBuilder();
	static int[][] graph;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());

		graph = new int[N][N];

		for (int i = 0; i < N; i++) {
			String row = br.readLine();
			for (int j = 0; j < N; j++) {
				graph[i][j] = Character.getNumericValue(row.charAt(j));
			}
		}
//		for (int i = 0; i < N; i++) {
//			System.out.println(Arrays.toString(graph[i]));
//		}
		quad_tree(0, 0, N);
		System.out.println(sb);
	}

	static void quad_tree(int x, int y, int size) {
		if (size == 1) {
			sb.append(graph[x][y]);
			return;
		}

		else {
			int sum = 0;
			for (int i = x; i < x + size; i++) {
				for (int j = y; j < y + size; j++) {
					sum += graph[i][j];
				}
			}
			if (sum % (size * size) == 0) {
				sb.append(graph[x][y]);
				return;
			} else {
				sb.append("(");
				quad_tree(x, y, size / 2);
				quad_tree(x, y + size / 2, size / 2);
				quad_tree(x + size / 2, y, size / 2);
				quad_tree(x + size / 2, y + size / 2, size / 2);
				sb.append(")");
			}

		}

	}

}